### PR TITLE
Reattach gateways

### PR DESCRIPTION
**Pull request by NicholasBlaskey**
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hamidzr — COMMENTED** [review]
the changes lgtm but I'm having and the tests pass on ci but some it looks like master isn't using the right port

```
NFO[2024-06-10T07:09:58-05:00] created gateway proxy listeners1 [1025]       component=kubernetesrm-worker id=3
DEBU[2024-06-10T07:09:58-05:00] launching service with spec {{ } {det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76-0  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[determined:e1e56d99-66ad-4791-8cf5-545f155dcd76.1] map[] [] [] []} {[{ TCP <nil> 3272 {0 0 } 0}] map[determined:e1e56d99-66ad-4791-8cf5-545f155dcd76.1]  [] ClusterIP []   []   0 false nil [] <nil> <nil> <nil> <nil>} {{[]} []}}  component=kubernetesrm-worker id=3
INFO[2024-06-10T07:09:58-05:00] pod is pulling images and starting            allocation-id=e1e56d99-66ad-4791-8cf5-545f155dcd76.1 component=job job=det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76 job-id=dc21a400-18a5-40f8-af78-2be0c7cdd7ff pod-name=det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76-z2n86 task-id=e1e56d99-66ad-4791-8cf5-545f155dcd76 task-type=SHELL
DEBU[2024-06-10T07:09:58-05:00] resources state changed: id=e1e56d99-66ad-4791-8cf5-545f155dcd76.1 state=PULLING reason=  job-id=dc21a400-18a5-40f8-af78-2be0c7cdd7ff task-id=e1e56d99-66ad-4791-8cf5-545f155dcd76 task-type=SHELL
DEBU[2024-06-10T07:09:58-05:00] resources state changed: id=e1e56d99-66ad-4791-8cf5-545f155dcd76.1 state=STARTING reason=  job-id=dc21a400-18a5-40f8-af78-2be0c7cdd7ff task-id=e1e56d99-66ad-4791-8cf5-545f155dcd76 task-type=SHELL
DEBU[2024-06-10T07:09:58-05:00] launching tcproute with spec {{ } {det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76-0  default    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[determined:e1e56d99-66ad-4791-8cf5-545f155dcd76.1] map[] [] [] []} {{[{<nil> <nil> 0x14000b7ead0 contour 0x14000b7eaf0 0x14001549054}]} [{[{{<nil> 0x14000b7eb00 det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76-0 <nil> 0x14001549060} <nil>}]}]} {{[]}}}  component=kubernetesrm-worker id=3
INFO[2024-06-10T07:09:58-05:00] created gateway proxy resources               component=kubernetesrm-worker id=3
DEBU[2024-06-10T07:10:00-05:00] starting pod log streamer                     podName=det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76-z2n86
INFO[2024-06-10T07:10:00-05:00] pod is running                                allocation-id=e1e56d99-66ad-4791-8cf5-545f155dcd76.1 component=job job=det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76 job-id=dc21a400-18a5-40f8-af78-2be0c7cdd7ff pod-name=det-941744f4-cmd-e1e56d99-66ad-4791-8cf5-545f155dcd76-z2n86 task-id=e1e56d99-66ad-4791-8cf5-545f155dcd76 task-type=SHELL
DEBU[2024-06-10T07:10:00-05:00] resources state changed: id=e1e56d99-66ad-4791-8cf5-545f155dcd76.1 state=RUNNING reason=  job-id=dc21a400-18a5-40f8-af78-2be0c7cdd7ff task-id=e1e56d99-66ad-4791-8cf5-545f155dcd76 task-type=SHELL
INFO[2024-06-10T07:10:00-05:00] registering service: e1e56d99-66ad-4791-8cf5-545f155dcd76 (http://127.0.0.1:1025)  component=proxy
DEBU[2024-06-10T07:10:00-05:00] registered proxy id: e1e56d99-66ad-4791-8cf5-545f155dcd76, cfg: &{e1e56d99-66ad-4791-8cf5-545f155dcd76 3272 true true}  job-id=dc21a400-18a5-40f8-af78-2be0c7cdd7ff task-id=e1e56d99-66ad-4791-8cf5-545f155dcd76 task-type=SHELL
INFO[2024-06-10T07:10:02-05:00] registering service: e1e56d99-66ad-4791-8cf5-545f155dcd76 (http://10.244.0.11:3272)  component=proxy
DEBU[2024-06-10T07:10:02-05:00] registered proxy id: e1e56d99-66ad-4791-8cf5-545f155dcd76, cfg: &{e1e56d99-66ad-4791-8cf5-545f155dcd76 3272 true true}  job-id=dc21a400-18a5-40f8-af78-2be0c7cdd7ff task-id=e1e56d99-66ad-4791-8cf5-545f155dcd76 task-type=SHELL
DEBU[2024-06-10T07:10:12-05:00] informer got new node event for node 'gw': MODIFIED   namespace=default

```
it's registering two services and end up using the second one? just switching to the target branch doesn't show this issue.

**Review by hamidzr — APPROVED** [review]
needed to use the new harness. 🚀

File: master/internal/rm/kubernetesrm/gateway_spec.go [hunk 1]
```diff
@@ -19,3 +23,33 @@ func createListenerForPod(gwPort int) gatewayTyped.Listener {
...
+func stripIndexFromSharedName(n string) string {
+	lastHyphenIndex := strings.LastIndex(n, "-")
+	if lastHyphenIndex == -1 {
+		return ""
+	}
+
+	return n[:lastHyphenIndex]
+}
+
+func generateListenerName(allocationID model.AllocationID, port int) string {
```
[the remark this answers is not in the document]
Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/gateway_spec.go:36]: https://hpe-aiatscale.atlassian.net/browse/RM-320

File: master/internal/rm/kubernetesrm/gateway_spec.go [hunk 2]
```diff
@@ -19,3 +23,33 @@ func createListenerForPod(gwPort int) gatewayTyped.Listener {
...
+	return fmt.Sprintf("%d-determined-%s", port, allocationID)
+}
+
+func listenerIsDetermined(listenerName string) bool {
+	return strings.Contains(listenerName, "determined")
+}
+
+func getAllocationIDFromListenerName(listenerName string) string {
```
[the remark this answers is not in the document]
Comment (NicholasBlaskey) [reply master/internal/rm/kubernetesrm/gateway_spec.go:44]: https://hpe-aiatscale.atlassian.net/browse/RM-320

**Pull request merged**